module/zfs/dnode.c

```diff
@@ -1519,22 +1532,22 @@ dnode_hold_impl(objset_t *os, uint64_t object, int flag, int slots,
 			return (SET_ERROR(EEXIST));
 		}
 
+		/* Don't actually hold if dry run, just return 0 */
+		if (flag & DNODE_DRY_RUN) {
```

Comment:
I use dry run, otherwise it will set slot to interiors and cause the next hold DNODE_MUST_BE_FREE to fail.

Reply:
Ahh, yes.  That makes sense, it's almost certainly why the existing code used `DNODE_MUST_BE_ALLOCATED` instead of `DNODE_MUST_BE_FREE`.